cibuildwheel/architecture.py

```diff
@@ -81,12 +81,14 @@ def all_archs(platform: PlatformName) -> "Set[Architecture]":
                 Architecture.ppc64le,
                 Architecture.s390x,
             }
-        elif platform == "macos":
+
+        if platform == "macos":
             return {Architecture.x86_64, Architecture.arm64, Architecture.universal2}
-        elif platform == "windows":
+
+        if platform == "windows":
             return {Architecture.x86, Architecture.AMD64, Architecture.ARM64}
-        else:
-            assert_never(platform)
+
+        assert_never(platform)
```

Comment:
> Could we disable this lint rule and restore these refactors?

We don't _have_ to have this lint rule enabled 😆. However, on reading large amounts of code and performing this refactor (cibuildwheel, plumbum, boost-histogram, and partially awkward), I find it to be very helpful for nested structures, and a tradeoff for shallow ones; and even then, it sometimes prompts real improvements, like moving the "return" clauses to the top. It reduces unnessiary whitespace, and it avoids dangling "else"'s that that are hard to match up with if's.

The key thing for me is you _can't_ ignore the body of the if - they have return's in them, and returns _are_ control flow, so the "else" is completely redundant and can be removed. Since it _can_ be removed, you have three choices: alway include it (terrible, would add tons of whitespace), always disallow it (this check), or mix and match and force the reader to have to digest both.

While when you write it, you might have an intent you are trying to convey, but there is no code impact at all; it is transparent to Python, so it really doesn't help the reader. It also creates long elif chains that are mentally taxing, while you can look at each guard clause in isolation; they are just "narrowing" (and this is exactly how MyPy works, actually).

Reply:
I really wish there was some sort of setting that would allow one layer, but not more. Many of the other checks have this sort of a knob, but I don't think these three do.

Reply:
> While when you write it, you might have an intent you are trying to convey, but there is no code impact at all; it is transparent to Python, so it really doesn't help the reader. 

Ah, see, well, I completely disagree with this statement, philosophically. We carefully name all our functions, variables, organise our code into classes and modules - this has no bearing on the ultimate working of the script, it could all be minified into one file and Python would execute it exactly the same. The code, the structure is for humans, to make it easier to understand!

> The key thing for me is you can't ignore the body of the if - they have return's in them, and returns are control flow, so the "else" is completely redundant and can be removed. 

Hmm, not sure I understand. If you're reading the above code while debugging a Windows issue, you don't need to read the mac section.

> Since it can be removed, you have three choices: alway include it (terrible, would add tons of whitespace), always disallow it (this check), or mix and match and force the reader to have to digest both.

Yeah, the third one. Except I'd phrase it as "stop generic rules from interfering with humans expressing ideas" :P

Reply:
> We carefully name all our functions

We have to choose a name, so we choose a name that conveys meaning. Here are not doing that, we are adding extra control structures that have no meaning and are completely ignorable.

Just curious, what about:

```python
for f in funcs:
    if f():
        return True
else:
    return False

```

? This is not hypothetical - this is an actual structure I'm clearing out of a different package.

Reply:
(note the alignment of the else - that was intentional)

Reply:
> Here are not doing that, we are adding extra control structures that have no meaning and are completely ignorable.

I'd argue that they do have meaning. They serve to express that this statement is meant to be read like a 'switch'.



> Just curious, what about:
> 
> ```python
> for f in funcs:
>     if f():
>         return True
> else:
>     return False
> ```

for...else has never made much sense to me. The analogy to if...else is weak, because we did execute the `for` already. My opinion is that the above is more readable without the `else`.

Reply:
So is:
```Python
if g():
    if f():
        return True
else:
    return False
```
better than:
```Python
if g():
    if f():
        return True
return False
```
? Imagine this much longer, of course, where aligning the else's is harder.

Reply:
Shouldn't we get rid of `if`/`elif`/`else` altogether for this one ? IMHO, the real intent is a map and maybe doing just that would be better for readability ?

```python
    @staticmethod
    def all_archs(platform: PlatformName) -> "Set[Architecture]":
        all_archs_map = {
            "linux": {
                Architecture.x86_64,
                Architecture.i686,
                Architecture.aarch64,
                Architecture.ppc64le,
                Architecture.s390x,
            },
            "macos": {Architecture.x86_64, Architecture.arm64, Architecture.universal2},
            "windows": {Architecture.x86, Architecture.AMD64, Architecture.ARM64},
        }
        return all_archs_map[platform]
```